Week 2 materials (#4)
NSLS2/bluesky-training-2025-2: Week2/README.md

# Hello Python and Jupyter 👋

## Setup

> 1. Open up : jupyter.nsls2.bnl.gov

Reviewer: ```suggestion
1. Open up: https://jupyter.nsls2.bnl.gov
```

2. Log in with your BNL credentials and accept DUO push
3. Start My Server
4. Select Job Profile: Scientific Python, Start
5. Open a notebook
6. Make a fork of this repository (if you have not done so already)
7. Copy the HTTP clone link
8. In the first cell of your notebook try:
```
!git clone https://github.com/<username>/bluesky-training-2025-2.git
```
9. You should see a new folder “bluesky-training-2025-2” appear in the file explorer

> 10. Open the folder and find “hello-python-and-jupyter.ipynb” in the week 2 folder

Reviewer: ```suggestion
10. Open the folder and find “hello-python-and-jupyter.ipynb” in the `Week2` folder
```

## Types of Jupyter Notebook Cells
**Markdown**
* Cells for text using Markdown which is like HTML (but easier)
* Headings, bold, italics, lists

> * https://daringfireball.net/projects/markdown/

Reviewer: Consider adding this link from Jupyter: https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html

**Code**
* Allows you to enter and run code
* Run each code cell using Shift-Enter or pressing the play button in the tool bar
* https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Running%20Code.html

**List of toolbar shortcuts:** https://gist.github.com/discdiver/9e00618756d120a8c9fa344ac1c375ac

## Other Python Interfaces

> * IPython – terminal based with magic by typing ```ipython```
> * Terminal – simple python shell by typing ```python``` 

Reviewer: If you need a code style inline, single back ticks are fine:
```suggestion
* IPython – terminal based with magic by typing `ipython`
* Terminal – simple python shell by typing `python`
```

* IDEs 
  * PyCharm - https://www.jetbrains.com/pycharm/

>   * Visual Studio Code - https://code.visualstudio.com

Reviewer: I'd put vscode as the first item as it's already installed on workstations.

## Python Environments

> **Conda** - Cross-platform, package manager 

Reviewer: ```suggestion
**Conda** - Cross-platform package manager 
```

* We use custom conda environments in DSSI to deploy necessary packages and python version on the experimental floor
* https://github.com/conda/conda

**venv** – virtual environments
* Module supports lightweight virtual environments each with their own independent set of python packages installed in their site directories
* Common install tool is pip
* https://docs.python.org/3/library/venv.html

**pixi** – new package management workspace
* https://pixi.sh/latest/

## Testing in Python

> * pytest: helps you write better programs

Reviewer: ```suggestion
* **pytest**: helps you write better programs
```
